### PR TITLE
Python reference pages

### LINKED ISSUE #28: Create docs page Reference/Python
### Purpose
Document Python conventions and libraries

### Page Tree
```yml
nav:
  ...
  - Reference:
    ...
    - Python: Reference/python.md
```

### Contents
- Conventions
    - PEP 8, flake8
        - File structure: imports, constants, main function, helper functions, `if __name__ == '__main__':`
    - camelcase vs underscores
- Virtual environments
    -  requirements.txt
        - explore https://github.com/bndr/pipreqs
    - [ ] Point virtual environment link in main README.md to this page

### Relevant Resources
- https://patrick-5546.github.io/notes/reference/python/

### PR DESCRIPTION
**Pull request by DFriend01**
Resolves #28. 

Created python reference pages that cover the following topics:

1. Python installation and learning resources
2. Virtual environments
3. Sailbot conventions for style guides and documentation in our source code

Be sure to check for grammar and spelling. If you see any incorrect information or missing information that you think that we should add, then please let me know.

### CODE REVIEW

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Since we don't really use python virtual environments, I would suggest changing the first sentence to "We recommend ... <because ...> but feel free to ..."

File: mkdocs.yml
Comment (patrick-5546) [inline mkdocs.yml]: I think conventions should come before virtual environment, whether it is because it's more relevant or because it's in alphabetical order

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: This brings up the question where do I want to create it in? Maybe simplify things by saying that you should run it inside the project that you want to create the virtual environment for

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: You could mention that the environment name is configurable but recommend `.venv` and use it for the commands afterwards so that they work when copy-pasted

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Would local be a better word than modular?

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Actually I think you can give the virtualenv init command the path. Also consider deleting this row because it is a little redundant with the row above.

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: I feel like this is more of an opinion than a difference

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: I think you should make this the first method you talk about because it is simpler

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Is there a website that you can go to search for conda packages and their channels? Or you could recommend googling the package's channel: "conda <PACKAGE> channel"

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Is it possible to make a nested tab for this?

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: Why not add the name argument to the example command above?

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md]: Not sure if the last phrase is necessary when before it you specify these conventions only apply when programming in Python

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md:11]: We should add some recommended readings for PEP 8, I like https://realpython.com/python-pep8/. Also we could state that the PEP 8 page you linked is a reference

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md]: Maybe "blocked from being merged" is more accurate than "rejected"

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md]: With the current/planned infrastructure there is no need for members to know how to install these or run them in the terminal: they just have to run the linting task in VS Code. If they are curious they can look in `.vscode/tasks.json` to see what command the tasks run.

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md]: We should use `"""` to follow PEP 8

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md]: I think the proper type is `List[float]`

File: docs/reference/python/conventions.md
Comment (patrick-5546) [inline docs/reference/python/conventions.md:59]: We should discuss whether we want to enforce using type hints. They work well with autoDocstring, but I feel like they are a bit restrictive

File: docs/reference/python/index.md
Comment (patrick-5546) [inline docs/reference/python/index.md]: I don't think we need this section. I can only imagine us recommending using sailbot_workspace, WSL Ubuntu, or native Ubuntu, which all comes with Python. For anything else it's the person's responsibility to figure out.

File: docs/reference/python/index.md
Comment (patrick-5546) [inline docs/reference/python/index.md]: You could change this sentence to something along the lines of "Feel free to add any resources that you found helpful to this list!"

**Review by patrick-5546 — CHANGES_REQUESTED** [review]
Great work, most of the comments I had were related to simplifying things for Python beginners

File: docs/reference/python/conventions.md
Comment (DFriend01) [reply docs/reference/python/conventions.md]: I think that standard is more for if you are autogenerating documentation.

File: docs/reference/python/conventions.md
Comment (DFriend01) [reply docs/reference/python/conventions.md:59]: We talked about this already. You said that we should avoid them. I originally thought that they were a good idea, but I am starting to change my mind because they are not the most convenient thing to use either. It requires you to know the type of the elements you are passing in, and that can get harder with custom classes or imported types.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: In my experience, you just gotta google it. I usually google "conda install \<package\>" and I usually get what I want.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: I thought they were not supported but I will try.

EDIT: It works.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: It's not needed. Adding the name is optional.

EDIT: Even if it is optional, it is best practice.

File: docs/reference/python/conventions.md
Comment (DFriend01) [reply docs/reference/python/conventions.md]: Will delete this section.

File: docs/reference/python/conventions.md
Comment (DFriend01) [reply docs/reference/python/conventions.md]: Will change this.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: When I say at the location, you need to give the path. For anaconda, it's in the environment variables so you don't need to remember the path. I'll change the wording.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: I don't understand this comment. What is `.venv`? What commands are we copy and pasting?

EDIT: `.venv` is hidden and is just a name for the folder just like `venv`.

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: I disagree. The whole point is that you use the conda installer. You want to use pip when conda does not work, hence the order that I put it in. It makes package resolution slightly easier in my experience.

EDIT: Just go with the easier one rather than the technically correct.

File: docs/reference/python/index.md
Comment (patrick-5546) [inline docs/reference/python/index.md]: @hhenry01 do you imagine that Network Systems will use Python for the new project? If not, we should update this to say that it is used by Pathfinding and Controls

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [inline docs/reference/python/virtual-environments.md]: You could make tabs for the different OS's

**Review by patrick-5546 — DISMISSED** [review]
LGTM, once you resolve the following comments you can complete this PR.

Once this is merged, you can update the Python virtual environment link in the README to https://ubcsailbot.github.io/docs/reference/python/virtual-environments/

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: Ok. I also think I wanna add some of those fancy icons in the tabs :)

I'll look into that

EDIT: Can't figure out how to use the simple icons resource to get the icons I want. @patrick-5546 maybe you know how to do this?

File: docs/reference/python/index.md
Comment (DFriend01) [reply docs/reference/python/index.md]: Do we really need to be specific here? Subteams might change in the future, and I feel like we can just direct new members which resources to consult on here.

File: docs/reference/python/virtual-environments.md
Comment (patrick-5546) [reply docs/reference/python/virtual-environments.md]: You can search for icons [here](https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/#search). I would recommend sticking with the included icons because adding custom ones are a pain

File: docs/reference/python/index.md
Comment (patrick-5546) [reply docs/reference/python/index.md]: I just wanted to get across that it might not be necessary for all software members to know python

File: docs/reference/python/virtual-environments.md
Comment (DFriend01) [reply docs/reference/python/virtual-environments.md]: Resolved by addressing dependency issues.

**Pull request merged**
